twisted/twisted/plugins/tunnel_helper_plugin.py

```diff
@@ -333,77 +401,107 @@ def cb(ds):
-                                                 '(only available if the crawler is enabled)', type=int)
-        parser.add_argument('-y', '--yappi', help="Profiling mode, either 'wall' or 'cpu'")
-        parser.add_help = True
-        args = parser.parse_args(sys.argv[1:])
-
-    except argparse.ArgumentError:
-        parser.print_help()
-        sys.exit(2)
-
-    socks5_port = int(args.socks5) if args.socks5 else None
-    introduce_port = int(args.introduce) if args.introduce else None
-    dispersy_port = int(args.dispersy) if args.dispersy else -1
-    crawl_keypair_filename = args.crawl
-    profile = args.yappi if args.yappi in ['wall', 'cpu'] else None
-
-    if profile:
-        yappi.set_clock_type(profile)
-        yappi.start(builtins=True)
-        logger.error("Profiling using %s time" % yappi.get_clock_type()['type'])
+class TunnelHelperServiceMaker(object):
+    implements(IServiceMaker, IPlugin)
+    tapname = "tunnel_helper"
+    description = "Tunnel Helper twistd plugin, starts a (hidden) tunnel as a service."
+    options = Options
 
-    if crawl_keypair_filename and not os.path.exists(crawl_keypair_filename):
-        logger.error("Could not find keypair filename", crawl_keypair_filename)
-        sys.exit(1)
+    def __init__(self):
+        """
+        Initialize the variables of this service and the logger.
+        """
+        self._stopping = False
+
+    def shutdown_process(self, shutdown_message, code=1):
+        msg(shutdown_message)
+        reactor.addSystemEventTrigger('after', 'shutdown', os._exit, code)
+        reactor.stop()
+
+    def start_tunnel(self, options):
+        """
+        Main method to startup a tunnel helper and add a signal handler.
+        """
+
+        socks5_port = options["socks5"]
+        introduce_port = options["introduce"]
+        dispersy_port = options["dispersy"]
+        crawl_keypair_filename = options["crawl"]
+        profile = options["yappi"]
+
+        if profile:
+            yappi.set_clock_type(profile)
+            yappi.start(builtins=True)
+            logger.error("Profiling using %s time" % yappi.get_clock_type()['type'])
+
+        settings = TunnelSettings()
+
+        # For disabling anonymous downloading, limiting download to hidden services only.
+        settings.min_circuits = 0
+        settings.max_circuits = 0
+
+        if socks5_port is not None:
+            settings.socks_listen_ports = range(socks5_port, socks5_port + 5)
+        else:
+            settings.socks_listen_ports = [random.randint(1000, 65535) for _ in range(5)]
+
+        settings.become_exitnode = options["exit"]
+        if settings.become_exitnode:
+            logger.info("Exit-node enabled")
+        else:
+            logger.info("Exit-node disabled")
+
+        tunnel = Tunnel(settings, crawl_keypair_filename, dispersy_port)
+        StandardIO(LineHandler(tunnel, profile))
```

Comment:
Yup, works. When I enter "p"  I get profiling disabled! with the logger. Speaking of profiling, would you mind running one of them with yappi enabled so I get a good profile output to analyse the tunnels with?